meja/typechecker.ml

```diff
@@ -0,0 +1,108 @@
+open Core_kernel
+open Parsetypes
+
+let rec check_type_aux typ ctyp env =
+  let without_instance ~f (typ : type_expr) env =
+    match Envi.Type.instance env typ with
+    | Some typ' -> (
+        let env = Envi.Type.clear_instance typ env in
+        let env = f typ' env in
+        match Envi.Type.instance env typ with
+        | Some _ ->
+            failwith
+              "Found a type variable that unifies with part of its own instance"
+        | None -> Some (Envi.Type.add_instance typ typ' env) )
+    | None -> None
+  in
+  match (typ.type_desc, ctyp.type_desc) with
+  | _, _ when Int.equal typ.type_id ctyp.type_id -> env
+  | Tvar (_, depth), Tvar (_, constr_depth) -> (
+    match without_instance typ env ~f:(fun typ -> check_type_aux typ ctyp) with
```

Comment:
This part would probably be a lot easier to read if you built a `bind_none` helper function (`val bind_none : 'a option -> f:(unit -> 'a option) -> 'a option`).